### PR TITLE
Create connection on background thread to avoid stalling the UI when loading VS

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/454212

There are two parts to this fix:

1. We get an OOP connection ready on the BG thread before syncing over on the UI thread.  This takes the cost of spinning up servicehub and whatnot off the UI thread.
2. We move to a single sync request to the OOP server instead of two requests.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @heejaechang 
May have to take for 15.3 But i wanted eyes on this first to see if people are ok with this approach.

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs [hunk 1]
```diff
@@ -18,6 +18,12 @@ private class WorkspaceHost : ForegroundThreadAffinitizedObject, IVisualStudioWo
             private readonly VisualStudioWorkspaceImpl _workspace;
             private readonly RemoteHostClient _client;
 
+            /// <summary>
+            /// A <see cref="RemoteHostClient.Connection"/> to use if non-null.  If null, we'll create 
+            /// our own connection to the OOP process.
+            /// </summary>
+            private Connection _preferredConnection;
```
Comment (sharwell) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs]: :bulb: Using `ReferenceCountedDisposable<Connection>` here would mean the caller wouldn't need to check a `dispose` field and apply conditional logic.

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs [hunk 2]
```diff
@@ -45,26 +57,44 @@ public void OnSolutionAdded(SolutionInfo solutionInfo)
...
-                        new object[] { solutionId, _workspace.DeferredState?.ProjectTracker.GetWorkingFolderPath(_workspace.CurrentSolution) },
-                        CancellationToken.None).ConfigureAwait(false);
+                        nameof(IRemoteHostService.RegisterPrimarySolutionId), 
+                        new object[] { solutionId, storageLocation }, CancellationToken.None).ConfigureAwait(false);
+                }
+                finally
+                {
+                    if (dispose)
+                    {
+                        connection?.Dispose();
```
Comment (sharwell) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs]: :bulb: No need for `?.` here; `connection` cannot be null.

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs [hunk 3]
```diff
@@ -74,17 +74,25 @@ internal sealed partial class ServiceHubRemoteHostClient : RemoteHostClient
...
+                // to UI thread to run. 
+                await Task.Factory.SafeStartNew(() =>
+                {
+                    host.SetPreferredConnection(connection);
+                    var projectTracker = vsWorkspace.GetProjectTrackerAndInitializeIfNecessary(Shell.ServiceProvider.GlobalProvider);
+
+                    projectTracker.RegisterWorkspaceHost(host);
+                    projectTracker.StartSendingEventsToWorkspaceHost(host);
+
+                    host.SetPreferredConnection(null);
```
Comment (sharwell) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs]: :bulb: This should be in a `finally` block

:question: Why are we not setting this to null before calling `RegisterWorkspaceHost`? It appears that the host is exposed to other code by that call, making it possible for unrelated code to use the preferred connection where it isn't expected to.

Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs]: > making it possible for unrelated code to use the preferred connection where it isn't expected to.

That's actually the point of this change :)

Right now we create the workspace host and then register it.  The project tracker then sends the host all the state changes it has tracked.  If any of these state changes need to be remoted to the OOP server, we have to expensively spin up a connection, and that impacts hte UI thread.  The point is to have the connection ready to go for this initial set of events to be sent over.  Once that happens, in teh future, we'll just acquire a connection on demand like we were doing before.

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs]: I'm not sure that's true.  The old code was written as if ```_client.TryCreateConnectionAsync``` could return null.  So i'm ensuring that if that does happen (either for creating the initial connect, or any subsequent connections) that we're resilient to it.

**Comment by CyrusNajmabadi** [comment]
> Need to verify that neither RegisterWorkspaceHost nor StartSendingEventsToWorkspaceHost can result in a race condition where _preferredConnection is used in the wrong location.

Sure.  What info do you need?  Right now the *intent* of the change is to make it so that those two calls *do* use preferred connection.  Note that all these workspace events are thread affinitized, and i only set the preferred connection on the UI thread, have all the events be sent, and then clear the preferred connection.  

Since no events can come in on a BG thread (and there are asserts in this code to track that), and all this work is done on a single UI thread call, i don't think _preferredConnection can be used in a wrong location.

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs]: >  This should be in a finally block

Yup!

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs]: Let me look into that

**Comment by sharwell** [comment]
> Sure. What info do you need?

I need to know that at the point the preferred connection is disposed (at the end of the `using` statement), it is not possible for code to be using the connection as part of an asynchronous request. The normal pattern for creating connections for use in a particular call naturally ensures this. For me, the most troublesome part of the change here is it removes the ability to visually inspect the scoping of the connection and the call.

**Comment by CyrusNajmabadi** [comment]
> it is not possible for code to be using the connection as part of an asynchronous request. 

Hopefully with RefCountedDisposable, that can't be a problem anymore.

**Comment by sharwell** [comment]
> Hopefully with RefCountedDisposable, that can't be a problem anymore.

Yes, it's looking like this is even more suitable to resolving this than I originally anticipated 👍

**Comment by CyrusNajmabadi** [comment]
Any more changes you'd like?
@heejaechang can you take a look?
@Pilchie I can spin up a vlaidation run for this against 15.3  However, changes like this worry me, and i'd prefer an exception along with a "we think we have a fix we can get in, but we'd like time to bake it since it hits core infrastructure".

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs [hunk 4]
```diff
@@ -45,26 +54,48 @@ public void OnSolutionAdded(SolutionInfo solutionInfo)
                 RegisterPrimarySolutionAsync().Wait();
             }
 
+            private async Task<ReferenceCountedDisposable<Connection>> GetConnectionAsync()
+            {
+                this.AssertIsForeground();
```
Comment (sharwell) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs]: ❔ Does this really need to be on the UI thread? I notice that `TryCreateConnectionAsync` uses `ConfigureAwait(false)` so we don't necessarily *stay* on the UI thread.

:bulb: If something here *does* need to be on the UI thread, why not use the equivalent of this?

```
await SwitchToMainThreadAsync();
```

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs [hunk 5]
```diff
@@ -45,26 +54,48 @@ public void OnSolutionAdded(SolutionInfo solutionInfo)
...
+                    connectionRef = new ReferenceCountedDisposable<Connection>(connection);
+                    _currentConnection = new ReferenceCountedDisposable<Connection>.WeakReference(connectionRef);
+                }
+
+                return connectionRef;
+            }
+
             private async Task RegisterPrimarySolutionAsync()
             {
+                this.AssertIsForeground();
```
Comment (sharwell) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs]: ❔ Does this really need to be on the UI thread?

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs [hunk 6]
```diff
@@ -26,11 +33,13 @@ private class WorkspaceHost : ForegroundThreadAffinitizedObject, IVisualStudioWo
 
             public WorkspaceHost(
                 VisualStudioWorkspaceImpl workspace,
-                RemoteHostClient client)
+                RemoteHostClient client,
+                ReferenceCountedDisposable<Connection> currentConnection)
```
Comment (sharwell) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs]: 💭 `initialConnection`?

**Comment by CyrusNajmabadi** [comment]
@heejaechang Another option is for us to just create a single connection on the BG thread and use that single connection for the lifetime of the workspace host.  The concern there is whether or not that's a problem if, say, the OOP server recycles itself.

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs [hunk 7]
```diff
@@ -74,17 +74,23 @@ internal sealed partial class ServiceHubRemoteHostClient : RemoteHostClient
                 return;
             }
 
-            // RegisterWorkspaceHost is required to be called from UI thread so push the code
-            // to UI thread to run. 
-            await Task.Factory.SafeStartNew(() =>
+            // Create a connection to the host in the BG to avoid taking the hit of loading 
+            // service hub on the UI thread.
+            var currentConnection = await client.TryCreateConnectionAsync(WellKnownRemoteHostServices.RemoteHostService, CancellationToken.None).ConfigureAwait(false);
```
Comment (sharwell) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs]: 💭 It would be nice for `TryCreateConnectionAsync` to return an RCD, but that's for a later change.

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs [hunk 8]
```diff
@@ -26,11 +33,13 @@ private class WorkspaceHost : ForegroundThreadAffinitizedObject, IVisualStudioWo
...
             {
                 _workspace = workspace;
                 _client = client;
                 _currentSolutionId = workspace.CurrentSolution.Id;
+                _currentConnection = new ReferenceCountedDisposable<Connection>.WeakReference(currentConnection);
```
Comment (sharwell) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs]: ❕ Need to handle the case where `currentConnection` is null, either here or in the code that calls this constructor

**Comment by CyrusNajmabadi** [comment]
@heejaechang Actually, is this a use-case for your new KeepAliveSession type?  Effectively, we'd like a long running connection to the OOP server, which can appropriately handle if things happen to the connection automagically.

Thoughts?

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs (hunk 8 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs]: good point.

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs (hunk 6 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs]: flipflopped on this.  I'm ok with the name as is.

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs [hunk 9]
```diff
@@ -74,17 +74,29 @@ internal sealed partial class ServiceHubRemoteHostClient : RemoteHostClient
...
 
-            // RegisterWorkspaceHost is required to be called from UI thread so push the code
-            // to UI thread to run. 
-            await Task.Factory.SafeStartNew(() =>
+            // Create a connection to the host in the BG to avoid taking the hit of loading service 
+            // hub on the UI thread.  We'll initially set its ref count to 1, and we will decrement 
+            // that ref-count at the end of the using block.  During this time though, when the 
+            // projectTracker is sending events, the workspace host can then use that connection 
+            // instead of having to expensively spin up a fresh one.
+            var currentConnection = await client.TryCreateConnectionAsync(WellKnownRemoteHostServices.RemoteHostService, CancellationToken.None).ConfigureAwait(false);
```
Comment (heejaechang) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs]: how about you just use KeepAliveSession?

http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/SymbolSearch/SymbolSearchUpdateEngineFactory.cs,24

create the session from BG and use TryInvokeAsync

**Comment by CyrusNajmabadi** [comment]
Closing in favor of https://github.com/dotnet/roslyn/pull/20755

**Comment by Pilchie** [comment]
@CyrusNajmabadi GitHub says this is merged?  Did the right thing happen here?

**Comment by CyrusNajmabadi** [comment]
Yes.  The other branch that was merged was branched off of this.  So technically this branch was merged in.
